File: Problems & Solutions/S/Shortest Path by Removing K walls/05 December 2022/Python/Shortest Path by Removing K walls.py
#User function Template for python3

from collections import deque 
import math
class Solution:
    def isSafe(self, r, c, R, C, vis, key):
        return (r >= 0) and (r < R) and (c >=0 ) and (c < C) and (not key in vis)
    def shotestPath(self, grid, n, m, k):
        if len(grid) == len(grid[0]) == 1:
            return 0
        
        start = (0, 0, grid[0][0], 0) # R, C, obstacles, turn
        
        visited = {}
        q = deque((start,))
        
        while q:
            r, c, ct, turn = q.popleft()
            nxt = [
                (r,c-1),
                (r-1,c),
                (r+1,c),
                (r,c+1),
            ]
            for r2, c2 in nxt: 
                if 0 <= r2 < n and 0 <= c2 < m:
                    ct2 = ct + grid[r2][c2]
                    
                    nn = (r2, c2, ct2, turn+1)
                    
                    if ct2 <= k and (r2, c2) == (n-1,m-1):
                        return turn+1
                    
                    kk = (r2,c2)
                    # Don't explore a path we already reached with fewer obstacles
                    if ct2 <= k and visited.get(kk, math.inf) > ct2:
                        visited[kk] = ct2
                        q.append(nn)
        
        return -1

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        n,m,k=map(int,input().split())
        mat=[]
        for i in range(n):
            row = list(map(int,input().split()))
            mat.append(row)
        
        ob = Solution()
        print(ob.shotestPath(mat,n,m,k))
# } Driver Code Ends

File: Problems & Solutions/S/Shortest Path by Removing K walls/05 December 2022/Java/Shortest Path by Removing K walls.java
//{ Driver Code Starts
//Initial Template for JAVA

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {
            
            String St[] = read.readLine().split(" ");
            
            int n = Integer.parseInt(St[0]);
            int m = Integer.parseInt(St[1]);
            int k = Integer.parseInt(St[2]);
            
            int[][] mat = new int[n][m];
            
            for(int i=0; i<n; i++)
            {
                String S[] = read.readLine().split(" ");
                for(int j=0; j<m; j++)
                    mat[i][j] = Integer.parseInt(S[j]);
            }

            Solution ob = new Solution();
            System.out.println(ob.shotestPath(mat,n,m,k));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution {
    class Node{
        int r,c,k,dist;
        Node(int r,int c,int k,int dist){this.r=r; this.c=c; this.k=k; this.dist=dist; }
    }
    int shotestPath(int[][] grid, int m, int n, int k) {
        // code here
        int directions[] = {0,1,0,-1,0};
        Queue<Node> q = new LinkedList<>();
        boolean vis[][][] = new boolean[m][n][k+1];
        vis[0][0][0]=true;
        q.add(new Node(0,0,0,0));
        while(!q.isEmpty()){
            Node front = q.poll();
            if(front.r==m-1 && front.c==n-1) return front.dist;
            for(int dir=0;dir<4;++dir){
                int newr=front.r+directions[dir], newc=front.c+directions[dir+1];
                if(newr>=0 && newr<m && newc>=0 && newc<n){
                    int newk=front.k+grid[newr][newc];
                    if(newk<=k && !vis[newr][newc][newk]){
                        vis[newr][newc][newk]=true;
                        q.add(new Node(newr,newc,newk,front.dist+1));
                    }
                }
            }
        }
        return -1;
    }
};